### PR TITLE
[RDY] pandoc filter to convert blockquotes to divs

### PR DESCRIPTION
**Pull request by aaren**
This filter converts all blockquote-with-attributes to
div-with-attributes, as discussed in issue #4.

A blockquote will be converted if
1. it begins with a header
2. that header has attributes
3. those attributes contain a single class
4. that class is one of ['objectives', 'callout', 'challenge']

For example, this is a valid blockquote:

```
> ## Callout time! {.callout}
> Let's do something
```

and it will be converted into this markdown:

```
<div class='callout'>
## Callout time!
Let's do something.
</div>
```

and the equivalent html.

I've integrated this into the `Makefile`, so `make preview` should show you
what it does.

TODO:
- [x] only allow attributes with one of the allowed classes

Things to consider:
- [x] Is the spec for a blockquote sufficiently tight / flexible?
- [x] Does the code do what it needs to / says?
- [x] Is this code maintainable? Does it need clearer documentation?
- [x] Do we pull in [pandocfilters](https://github.com/jgm/pandocfilters) as a dependency?

### CODE REVIEW

**Comment by embray** [comment]
Is this just assigned to me to review, or...?

**Comment by aaren** [comment]
@embray I'm guessing review as last thing greg said to me was that he'd find a reviewer.

Let me know if anything needs clarifying.

**Comment by embray** [comment]
Okay sounds good--I'll get back to this shortly.

**Comment by embray** [comment]
Is there a reason to turn it into a `<div>` and not just a `<blockquote>`  with different styling?  At the end of the day they're still just block elements and the only thing that distinguishes them is the style defaults IIRC.

**Comment by embray** [comment]
Looks fine otherwise; I've never worked directly with pandoc before so I can't claim expertise but the code looks solid, and seems to do what it advertises.  I just don't fully understand the rationale for converting from a `<blockquote>` to a `<div>` (I wonder if other HTML-5 semantic elements might not be appropriate, like `<aside>` or `<section>`, etc.)

**Comment by gvwilson** [comment]
Agreed on the styling, but the semantics are (meant to be) different.

**Comment by gvwilson** [comment]
We're converting to 'div' because most of Bootstrap is written in terms 
of div's.

**Comment by aaren** [comment]
Pandoc has an internal concept of a Div, which is a generic container
with attributes. There is currently no markdown implementation of this
beyond explicitly using `<div>` tags.

Pandoc does have a concept of a Blockquote though, which is just another
container, and a markdown implementation of it using `>`. Blockquotes
don't have attributes, but that is what this PR gets around.

This provides a convenient way to indicate arbitrary content that we
want to have particular classes applied to it (css styling).

There might be a more semantic html5 way to do it but we would then have
to worry about browser support, which I don't know much about. However,
without too much effort, it would be possible to use any tag you like by
extending this filter. I'm happy to do that if it's desirable.

**Comment by aaren** [comment]
> Is there a reason to turn it into a `<div>` and not just a
> `<blockquote>`  with different styling?  

Yes, in pandoc we can't say what styling the blockquote should have. In
html it just comes out as `<blockquote>...</blockquote>`. So if you want
different styling for callouts, challenges, etc. then you need to
convert to a container on which we can set these, i.e. a div.

**Comment by aaren** [comment]
@embray @gvwilson what do you think about the specification for the class-able blockquotes? Is it too restrictive having to start with a header? This is the simplest way to implement it.

**Comment by gvwilson** [comment]
+1 to having it start with a header - easy for people who are only 
writing Markdown to recognize & create.

**Comment by aaren** [comment]
last thing: what about the [pandocfilters](https://github.com/jgm/pandocfilters) dependency? If users are using anaconda then we can install using pip. Alternately we can include the source (a single file) in the repo.

**Comment by gvwilson** [comment]
Please add a mention to the setup instructions - if you look in 
workshop-template's README.md, it opens with a description of how to set 
up dependencies, and we should model lesson-template's README.md after that.

**Comment by aaren** [comment]
I've put instructions in the README. I included an instruction to install pandoc and just linked to the pandoc site instructions. We might need to explain this a bit more but I think that is the subject of another PR (along with restructuring the readme).

**Comment by gvwilson** [comment]
What's the PR number?

File: README.md [hunk 1]
```diff
@@ -15,6 +15,17 @@ create a new lesson:
...
+
+1. Install Pandoc (see http://www.johnmacfarlane.net/pandoc/installing.html)
+
+2. Install pandocfilters, a python module that helps with writing
+   filters for pandoc:
+
+    ```
+    pip install pandocfilters
+    ```
+
```
Comment (gvwilson) [inline README.md:28]: Can pandocfilters also be installed using conda and other tools?

Comment (aaren) [inline README.md:28]: Don't think so - I can't find it on conda. Is this a problem?

**Comment by aaren** [comment]
This is PR #9 : https://github.com/swcarpentry/lesson-template/pull/9

File: README.md (hunk 1 above)
Comment (gvwilson) [inline README.md:28]: Not a problem, just want to give install instructions for all platforms.

Comment (aaren) [inline README.md:28]: I thought pip was multi-platform? I have it on windows.

Comment (gvwilson) [inline README.md:28]: Yup, but if people got their Python via Anaconda, we may want to tell 
them how to install with conda. Not a bit deal...

Comment (aaren) [inline README.md:28]: From the [anaconda FAQ](http://docs.continuum.io/anaconda/faq.html#install-packages):

> How do I install additional packages into Anaconda?
>     First, you may try to install the package using the conda command, i.e. try typing conda install <pkg > name. If the package is not available, you can install it by the usual means, e.g. pip...

So I think `pip` is the Anaconda way to install things that aren't available with `conda install`. Anaconda includes pip 1.5.6 by default.

Comment (gvwilson) [inline README.md:28]: Good enough - I think we can go ahead and merge this one.

File: README.md [hunk 2]
```diff
@@ -15,6 +15,17 @@ create a new lesson:
...
 9.  Tell us where your lesson is so that we can use it and help you improve it.
 
+### Dependencies
+
+1. Install Pandoc (see http://www.johnmacfarlane.net/pandoc/installing.html)
+
+2. Install pandocfilters, a python module that helps with writing
+   filters for pandoc:
+
+    ```
```
Comment (gvwilson) [inline README.md]: We use triple tildes for code blocks (I know that triple backquotes work as well, but we're trying to be consistent).

File: README.md [hunk 3]
```diff
@@ -15,6 +15,17 @@ create a new lesson:
 8.  Push your changes to GitHub for viewing.
 9.  Tell us where your lesson is so that we can use it and help you improve it.
 
+### Dependencies
+
+1. Install Pandoc (see http://www.johnmacfarlane.net/pandoc/installing.html)
```
Comment (gvwilson) [inline README.md]: Or `[Install Pandoc](http://www.johnmacfarlane.net/pandoc/installing.html)` ?

**Comment by aaren** [comment]
@gvwilson - changed backticks to tildes and added a little more doc in the filter. I think it's all good now :)

**Comment by gvwilson** [comment]
Thanks much.

**Pull request merged**